packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+}
+
+describe("DeviceSource", () => {
+    it("should exist", () => {
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        expect(mouseSource).not.toBeNull();
+        expect(mouseSource.deviceType).toBe(DeviceType.Mouse);
+        expect(mouseSource.deviceSlot).toBe(0);
+        expect(mouseSource.onInputChangedObservable).not.toBeUndefined();
+    });
+    it("can poll with getInput", () => {
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        const leftClick = mouseSource.getInput(PointerInput.LeftClick);
+        expect(leftClick).toBe(1);
+    });
+    it("can use onInputChangedObservable", () => {
+        expect.assertions(2);
+        let resultEvent: Nullable<IWheelEvent | IPointerEvent> = null;
+        let onInputChanged = null;
+        // Create DeviceSource and add to its observable
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, onInputChanged);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        onInputChanged = (deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent) => {
+            mouseSource.onInputChangedObservable.notifyObservers(eventData as IPointerEvent);
+        };
+        mouseSource.onInputChangedObservable.add((eventData) => {
+            resultEvent = eventData;
+        });
+
+        expect(mouseSource.onInputChangedObservable.hasObservers()).toBe(true);
+        const mouseEvent: IUIEvent = { pointerId: 0, type: "pointerdown", pointerType: "mouse", clientX: 1, clientY: 1 } as IPointerEvent;
+        onInputChanged(DeviceType.Mouse, 0, mouseEvent);
+
+        expect(resultEvent).toEqual(mouseEvent);
+    });
+});
+
+describe("DeviceSourceManager", () => {
+    let engine: Nullable<NullEngine> = null;
+
+    beforeEach(function () {
+        engine = new NullEngine();
+        engine._deviceSourceManager = new MockInternalDeviceSourceManager() as InternalDeviceSourceManager;
+    });
+
+    afterEach(function () {
+        engine?.dispose();
+    });
+
+    it("should exist", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        expect(deviceSourceManager).not.toBe(null);
+        expect(engine?._deviceSourceManager?._refCount).toBe(1);
+    });
+
+    it("can use getDeviceSource", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const nullSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 0);
+        expect(nullSource).toBe(null);

```

Comment:
nit: Maybe put a newline before this to separate it from the top section

Reply:
I'll separate and clean up this section2